### ISSUE TITLE
Offset value of the NF4 datatype

### ISSUE THREAD

**Issue by mscherrmann** [bitsandbytes-foundation/bitsandbytes#938]
I have a question about the implementation of the NF4 data type:

In the implementation process, I noticed that eight evenly spaced intervals between 0.5 and an offset value of 0.9677083 are created for the negative part, and similarly nine intervals for the positive part. I understand the general approach, but I'm unclear about the rationale behind the specific offset value. Is it chosen arbitrarily, or is there a specific reason for selecting this value? Initially, I thought the offset might be calculated as 2^k/(2^k+1) = 16/17 = 0.9412, based on the paper's guidance, but this seems to differ from the offset value you've used.

Could you please clarify this for me? Thank you very much!

**Comment by chenxiaodongman** [comment]
C:\Users\chenx>python -m bitsandbytes

===================================BUG REPORT===================================
Welcome to bitsandbytes. For bug reports, please submit your error trace to: https://github.com/TimDettmers/bitsandbytes/issues
================================================================================
binary_path: C:\Python312\lib\site-packages\bitsandbytes\cuda_setup\libbitsandbytes_cuda116.dll
CUDA SETUP: Loading binary C:\Python312\lib\site-packages\bitsandbytes\cuda_setup\libbitsandbytes_cuda116.dll...
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
++++++++++++++++++++ DEBUG INFORMATION +++++++++++++++++++++
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

++++++++++ POTENTIALLY LIBRARY-PATH-LIKE ENV VARS ++++++++++
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

WARNING: Please be sure to sanitize sensible info from any such env vars!

++++++++++++++++++++++++++ OTHER +++++++++++++++++++++++++++
COMPILED_WITH_CUDA = True
Traceback (most recent call last):
  File "C:\Python312\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Python312\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Python312\lib\site-packages\bitsandbytes\__main__.py", line 47, in <module>
    print(f"COMPUTE_CAPABILITIES_PER_GPU = {get_compute_capabilities(cuda)}")
  File "C:\Python312\lib\site-packages\bitsandbytes\cuda_setup\main.py", line 336, in get_compute_capabilities
    check_cuda_result(cuda, cuda.cuDeviceGetCount(ct.byref(nGpus)))
AttributeError: 'NoneType' object has no attribute 'cuDeviceGetCount'

C:\Users\chenx>

**Comment by FinTexIFB** [comment]
Sorry, but I don't understand your answer. Could you be a bit more precise? Thank you!

**Comment by TimDettmers** [comment]
Closing this issue. The offset value `0.9677083` in `create_normal_map` determines the outermost quantile boundary for the NF4 data type. It is chosen so that `scipy.stats.norm.ppf(0.9677083) ≈ 1.845`, meaning the outermost quantization bin covers values up to ~1.845 standard deviations from the mean. This value was empirically optimized to minimize quantization error for normally-distributed weights, rather than derived from the simpler formula `2^k/(2^k+1) = 16/17 ≈ 0.9412` suggested in the paper.

We are improving the documentation for these functions.

**Comment by TimDettmers** [comment]
The `create_normal_map` function and the NF4 offset parameter are now documented in commit d7c179c. The docstring explains how the offset controls the outermost quantile boundary and how the NF4 lookup table is constructed.
